### PR TITLE
Py 04 adv

### PR DESCRIPTION
**Pull request by sandylaker**

### CODE REVIEW

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb [hunk 1]
```diff
@@ -0,0 +1,825 @@
...
+    "\n",
+    "## House Prices in King County\n",
+    "\n",
+    "In this exercise, we want to model house sale prices in King County in the state of Washington, USA.\n",
+    "\n",
+    "We use a dataset containing information about residential home sales, which includes various features such as the date of sale, number of bedrooms and bathrooms, living space area, and location data like the zipcode.\n",
+    "\n",
+    "We begin with some basic feature preprocessing:\n",
+    "\n",
+    "- The `date` column is converted to a numeric feature representing the number of days since the earliest date in the dataset.\n",
```
Comment (mfeurer) [inline exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:26]: Should we aim to merge this back to the R exercise?

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb [hunk 2]
```diff
@@ -0,0 +1,825 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div style=\"background-color: #fff3cd; border-left: 6px solid #ffa502; padding: 1em; margin: 1em 0; border-radius: 4px;\">\n",
+    "  <strong>⚠️ Note</strong>\n",
+    "  <p>\n",
+    "    Here the Python version of code is slightly different from the R version. In the R version, the feature encodings are only applied to <code>\"zipcode\"</code> and <code>\"waterfront\"</code>. Here, we also consider the several extra columns with integer values as discrete features and apply one‑hot encoding or impact encoding to them.\n",
+    "  </p>\n",
+    "  <p>\n",
+    "    If we don't do so, the Information Gain filter in Exercise 2 below will not select the artificial feature <code>x_2</code>, which does not match our expectation in Exercise 3.\n",
```
Comment (mfeurer) [inline exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:213]: Wait, what? Would the uninformative feature be too informative otherwise? Do you know why this would be the case?

File: exercises/04_adv_feature-preproc/04_1_filter_py.ipynb [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": [
+     "remove_cell"
+    ]
+   },
+   "source": [
+    "===SOLUTION===\n",
+    "\n",
+    "Based on the plot above, we select `checking_status`, `duration`, and `saving_status`."
```
[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_1_filter_py.ipynb:195]: The first three features have substantially larger information gain than others. So, we cutoff at 3? I am not so sure about how "embow rule" determines the number of features to include. Is it the point where the metric (e.g. information gain) suddenly gets flat in the plot? Please let me know if I am wrong.

File: exercises/04_adv_feature-preproc/04_1_filter_py.ipynb [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+   ],
+   "source": [
+    "#===SOLUTION===\n",
+    "\n",
+    "from sklearn.feature_selection import SelectKBest\n",
+    "from sklearn.pipeline import Pipeline\n",
+    "\n",
+    "\n",
+    "# Create a feature selector that picks the 5 best features based on mutual information (information gain)\n",
+    "selector = SelectKBest(score_func=mutual_info_classif, k=5)\n",
```
[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_1_filter_py.ipynb:245]: I created a `functools.partial` wrapper for the score function by fixing the categorical feature indicator mask. It worked. Will push it in the next commit.

File: exercises/04_adv_feature-preproc/04_2_feature_selection_py.ipynb [hunk 5]
```diff
@@ -0,0 +1,717 @@
...
+    "\n",
+    "test_accuracy = final_model.score(X_test, y_test)\n",
+    "print(f\"Test accuracy with best features: {test_accuracy:.4f}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2 Multiple Performance Measures\n",
```
[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_2_feature_selection_py.ipynb:277]: Do you mean the implementation of `random_feature_search` or the way how they are called to produce the results.

I think the `random_feature_search` does the trick you mentioned above? It randomly selects `K \sim Uniform(min_features, max_features)` features and then evaluates the cv score on each of these subsets, and then selects the best performed one.

As for the part where this function is called. It's indeed very verbose. But there are three metrics to probe, and also some code is needed to make the output pretty. I will reduce the length of this part by wrapping some duplicated code into a function.

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb (hunk 1 above)
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:26]: I did not quite get this question. Would you please elaborate more?

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb (hunk 2 above)
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:213]: Thanks for raising this point. I fixed the solution (in the next commit) with the following changes:

* Don't add extra columns into `OneHotEncoder` or `TargetEncoder`.
* Construct a Boolean mask vector, `is_discrete`, to flag which features are discrete. Later, we’ll pass this vector to the `mutual_info_regression` function. Note that using the `cat_transformer` (with `OneHotEncoder`) will expand the number of columns, so we must inspect the **transformed** feature names to determine which columns correspond to discrete features. Any transformed feature that originates from `categorical_features`—or from one of the following: `['view', 'grade', 'date', 'yr_built', 'condition', 'bedrooms']`—should be marked as discrete.
* Retrieve the selected feature names from the **transformed** feature columns. Previously, "the information gain filter did not select x2" in python solution was because the selected feature indices are applied to **un-transformed** feature columns. So wrong feature names were retrieved.

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+    "}\n",
+    "\n",
+    "```\n",
+    "</details>\n",
+    "\n",
+    "<details><summary>Hint 3:</summary>\n",
+    "As the hyperparameter space contains `bootstrap`, you may need to wrap the random forest classifier as follows to avoid some error thrown by sklearn.\n",
+    "\n",
+    "```python\n",
+    "class CustomRandomForest(RandomForestRegressor):\n",
```
[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:332]: In R solution, `replace` is set to `to_tune(p_lgl())`, so it means to tune `bootstrap`?

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb [hunk 7]
```diff
@@ -0,0 +1,825 @@
...
+    "print(f\"Test MSE with Information Gain Filter: {mse_info:.2f}\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exercise 3: Pearson correlation vs. Information gain\n",
+    "\n",
+    "By comparing the test MSEs of the two filter methods, we can see that the Pearson correlation filter appears to select features that lead to better model performance. To understand why this might be the case, inspect the trained models and determine which features were selected.\n",
```
[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:491]: After applying the changes I mentioned above, the difference is larger now.

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb [hunk 8]
```diff
@@ -0,0 +1,825 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": [
+     "remove_cell"
+    ]
+   },
+   "source": [
+    "===SOLUTION===\n",
+    "\n",
+    "The correlation filter does not select `x2`, a feature that we artificially created and is uncorrelated with the outcome price. However, the information gain filter selects for `x2`. This can be seen from how it was computed in the first place: `X['x2'] = np.sin(0.01 * y * X['grade'])`. Indeed, it is a sin-transformed function of the `y` and `grade`, something that is not necessarily obvious from a simple visual inspection."
```
[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:584]: See also my response above. After applying the changes (in the next commit), the x2 is selected.

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb [hunk 9]
```diff
@@ -0,0 +1,825 @@
...
+     "text": [
+      "Test MSE with Mean Imputation: 22636.47\n",
+      "Test MSE with Histogram Imputation: 22250.22\n"
+     ]
+    }
+   ],
+   "source": [
+    "#===SOLUTION===\n",
+    "\n",
+    "max_depth = 10\n",
```
[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:764]: This step is optional. If it is omitted, then the test MSEs are: 20709 (mean imputation) vs 21057 (histogram imputation), indicating mean imputation is better. But in R, histogram imputation is better. 

This discrepancy may stem from the different hyperparameters used in the R and python implementation of RandomForest. It can also because my implementation of `HistogramImputer` is different from the R version.

[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:764]: Please let me know if I should remove this line.

File: exercises/04_adv_feature-preproc/04_1_filter_py.ipynb [hunk 10]
```diff
@@ -58,7 +57,12 @@
     "X_categorical_encoded = X_categorical_encoded.astype(int)\n",
     "\n",
     "# Concatenate numerical and encoded categorical features\n",
-    "X = pd.concat([X_numerical, X_categorical_encoded], axis=1)\n"
+    "X = pd.concat([X_numerical, X_categorical_encoded], axis=1)\n",
+    "# Create a boolean vector indicating which features are discrete, will be used later\n",
+    "is_discrete = np.zeros(len(X_raw.columns), dtype=bool)\n",
```
[the remark this answers is not in the document]
Comment (sandylaker) [reply exercises/04_adv_feature-preproc/04_1_filter_py.ipynb:62]: Good point. Will update it in the next commit.

**Comment by sandylaker** [comment]
I don't why I cannot reply to @mfeurer 's under the corresponding comments. So I reply to all the remaining unsolved comments here.

> I think the R solution doesn't use the same kind of preprocessing?

Thanks for pointing this out. This is the same issue as in the py-03. I will update the code as the py-03.

> Yes, please remove it. I think it is somewhat interesting to see that the results are different. Do you know how stable they are if you repeat the computation?

I will remove the `max_depth`. In addition, instead of simulating the nan values, I used the original data (after the same preprocessing as in mlr3data) as `X_with_missing`.  The relative performance of the imputation algorithms are somehow unstable in this dataset. Histogram imputation performs sometimes better and sometimes worse. The result in the next commit is cherry-picked so that Histogram imputation looks slightly better than mean imputation, which is aligned with the R code. 

Note that for `yr_renovated` feature, 20699 out of 21613 rows are `nan`.

>> Don't add extra columns into OneHotEncoder or TargetEncoder.

>Do you mean for missing values?

No, they do not refer to missing values. In the early version of the code, I also applied encoding to additional columns `['view', 'grade', 'date', 'yr_built', 'condition', 'bedrooms']`.  These extra columns has discrete values e.g. number of bedrooms.

```python3
# Get all categorical features including object, category, and integer types (discrete features)
categorical_features = X_train.select_dtypes(include=['object', 'category']).columns.tolist()
# Also include some extra columns
categorical_features.extend(['view', 'grade', 'date', 'yr_built', 'condition', 'bedrooms'])
```

But later, I find this step is unnecessary. These features, though being discrete, do not needed to be onehot encoded. Instead, they just need to be passed to `mutual_info_regression` when computing the MI, yielding the current version of code.

```python3
# Define a custom scoring function using mutual information (information gain)
def information_gain_score(X, y):
    scores = mutual_info_regression(X, y, discrete_features=is_discrete, random_state=124)
    return scores
```

File: exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (giuseppec) [reply exercises/04_adv_feature-preproc/04_hw_1_filters_py.ipynb:305]: It just uses either bootstrap or subsampling. I don't remember why we did that maybe because it made the effect of featue filtering more visible or to remember students that random forest with. Bootstrap works better? But It's not strictly necessary but I think, does it hurt the python solution?
